readme.md
Map
=======

Demo
-------

[Demo of the Map control](http://demo.askia.com/WebProd/cgi-bin/AskiaExt.dll?Action=StartSurvey&SurveyName=ADC2_Map)

Tested in
-----------

    This control was successfully tested in Chrome, Firefox, Safari, Opera and IE8+.

Dependencies
-----------------

    * Raphaël 2.1.4 - JavaScript Vector Library
    * Licensed under the MIT (http://raphaeljs.com/license.html) license

Notes
-------

	- Maps (France, United Kingdom, USA, Germany, Belgium or Italy)
    - Some parameters can have different rendering depending of the browser used.
    - The size of the map is fixed on IE8.
    - To create a line break in the caption, use backslash + n "\n"

![screenshot](/demo/adc2-map.png)


changelog.md
v1.0.0 - 2016/03/31 - Jérôme Duparc - first version of the control
v2.0.0 - 2017/02/16 - Jérôme Duparc - Theme implementation
v2.0.4 - 2018/02/22 - Jérôme Duparc - Switzerland implementation

resources/static/map-min.js
!function(){function a(a,b,c){"function"==typeof a.addEventListener?a.addEventListener(b,c,!1):a.attachEvent&&(a["e"+b+c]=c,a[b+c]=function(){a["e"+b+c].call(a,window.event)},a.attachEvent("on"+b,a[b+c]))}function b(a,b,c,d){document.querySelector("."+this.data("name")).checked||"area"!==this.data("strID")||a.forEach(function(a){return function(e){a===e.data("name")&&"path"===e.data("strID")?(e.attr({fill:b.fill,stroke:b.stroke,"stroke-width":b["stroke-width"],opacity:b.opacity,cursor:"pointer"}),!1===ie8&&"yes"===d&&e.animate({fill:b.fill,stroke:b.stroke,"stroke-width":b["stroke-width"],opacity:b.opacity,transform:"s1.03 1.03"},200)):a===e.data("name")&&"txt"===e.data("strID")&&e.attr({fill:c.fill,stroke:c.stroke,"stroke-width":c["stroke-width"],opacity:c.opacity,cursor:"pointer"})}}(this.data("name")))}function c(a,b,c,d){document.querySelector("."+this.data("name")).checked||"area"!==this.data("strID")||a.forEach(function(a){return function(e){a===e.data("name")&&"path"===e.data("strID")?(!1===ie8&&"yes"===d&&e.animate({fill:b.fill,stroke:b.stroke,"stroke-width":b["stroke-width"],opacity:b.opacity,transform:"s1.0 1.0"},200),e.attr({fill:b.fill,stroke:b.stroke,"stroke-width":b["stroke-width"],opacity:b.opacity,cursor:"pointer"})):a===e.data("name")&&"txt"===e.data("strID")&&e.attr({fill:c.fill,stroke:c.stroke,"stroke-width":c["stroke-width"],opacity:c.opacity,cursor:"pointer"})}}(this.data("name")))}function d(a,b,c,d,e,f,g){"area"===this.data("strID")&&(document.querySelector("."+this.data("name")).checked=!0,a.forEach(function(a){return function(g){a!==g.data("name")&&"path"===g.data("strID")?g.attr({fill:d.fill,stroke:d.stroke,"stroke-width":d["stroke-width"],opacity:d.opacity,cursor:"pointer"}):a!==g.data("name")&&"txt"===g.data("strID")?g.attr({fill:e.fill,stroke:e.stroke,"stroke-width":e["stroke-width"],opacity:e.opacity,cursor:"pointer"}):a===g.data("name")&&"path"===g.data("strID")?(g.attr({fill:b.fill,stroke:b.stroke,"stroke-width":b["stroke-width"],opacity:b.opacity,cursor:"pointer"}),!1===ie8&&"yes"===f&&g.animate({fill:b.fill,stroke:b.stroke,"stroke-width":b["stroke-width"],opacity:b.opacity,transform:"s1.0 1.0"},200)):a===g.data("name")&&"txt"===g.data("strID")&&g.attr({fill:c.fill,stroke:c.stroke,"stroke-width":c["stroke-width"],opacity:c.opacity,cursor:"pointer"})}}(this.data("name"))),window.askia&&window.arrLiveRoutingShortcut&&window.arrLiveRoutingShortcut.length>0&&window.arrLiveRoutingShortcut.indexOf(g)>=0&&askia.triggerAnswer())}function e(e){this.options=e,this.instanceId=e.instanceId||1,this.mapId=e.mapId||1,this.width=e.width||300,this.height=e.height||300,this.fontSize=e.fontSize||16,this.animatePath=e.animatePath||"yes",this.fill=e.fill||"#F7F7F7",this.strokeColor=e.strokeColor||"#A1A7AD",this.strokeWidth=e.strokeWidth||1,this.opacity=e.opacity||0,this.fillHover=e.fillHover||"#FFFFFF",this.strokeColorHover=e.strokeColorHover||"#A1A7AD",this.strokeWidthHover=e.strokeWidthHover||1,this.opacityHover=e.opacityHover||0,this.fillSelected=e.fillSelected||"#E8F1FF",this.strokeColorSelected=e.strokeColorSelected||"#A1A7AD",this.strokeWidthSelected=e.strokeWidthSelected||1,this.opacitySelected=e.opacitySelected||0,this.fillText=e.fillText||"#444444",this.opacityText=e.opacityText||0,this.fillTextHover=e.fillTextHover||"#444444",this.opacityTextHover=e.opacityTextHover||0,this.fillTextSelected=e.fillTextSelected||"#007BC6",this.opacityTextSelected=e.opacityTextSelected||0,this.adjustXtextObj=e.adjustXtextObj||[],this.adjustYtextObj=e.adjustYtextObj||[],this.mapCaptions=e.mapCaptions||[],this.mapNames=e.mapNames||[],this.mapPaths=e.mapPaths||[];for(var f=e.currentQuestion||"",g=this.animatePath,h={fill:this.fill,stroke:this.strokeColor,"stroke-width":this.strokeWidth,opacity:0,cursor:"pointer"},i={"stroke-linecap":"round","stroke-linejoin":"round",fill:this.fill,stroke:this.strokeColor,"stroke-width":this.strokeWidth,opacity:this.opacity,cursor:"pointer"},j={"stroke-linecap":"round","stroke-linejoin":"round",fill:this.fillHover,stroke:this.strokeColorHover,"stroke-width":this.strokeWidthHover,opacity:this.opacityHover},k={"stroke-linecap":"round","stroke-linejoin":"round",fill:this.fillSelected,stroke:this.strokeColorSelected,"stroke-width":this.strokeWidthSelected,opacity:this.opacitySelected},m={"font-size":this.fontSize,fill:this.fillText,stroke:"none","stroke-width":0,opacity:this.opacityText,"font-weight":"normal",cursor:"pointer"},n={"font-size":this.fontSize,fill:this.fillTextHover,stroke:"none","stroke-width":0,opacity:this.opacityTextHover,"font-weight":"normal"},o={"font-size":this.fontSize,fill:this.fillTextSelected,stroke:"none","stroke-width":0,opacity:this.opacityTextSelected,"font-weight":"normal"},p=Raphael("adc_"+this.instanceId+"_container"),q=p.set(),r=0;l=this.mapPaths.length,r<l;r++){var s=p.path(this.mapPaths[r]).attr(document.querySelector("."+this.mapNames[r]).checked?k:i).data("name",this.mapNames[r]).data("strID","path");q.push(s)}for(var r=0;l=this.mapCaptions.length,r<l;r++){var t=q[r].getBBox(),u=s.paper.text(t.x+t.width/2+(void 0===this.adjustXtextObj[r]?0:this.adjustXtextObj[r]),t.y+t.height/2+(void 0===this.adjustYtextObj[r]?0:this.adjustYtextObj[r]),this.mapCaptions[r]).attr(document.querySelector("."+this.mapNames[r]).checked?o:m).data("name",this.mapNames[r]).data("strID","txt");q.push(u)}for(var r=0;l=this.mapPaths.length,r<l;r++){var v=p.path(this.mapPaths[r]).attr(h).data("name",this.mapNames[r]).data("strID","area");q.push(v)}q.hover(function(){b.call(this,q,j,n,g)},function(){c.call(this,q,i,m,g)}),q.click(function(){d.call(this,q,k,o,i,m,g,f)}),p.setViewBox(0,0,this.width,this.height,!1),this.resizePaper(p,this.width,this.height),a(document.getElementsByTagName("BODY")[0],"resize",this.resizePaper(p,this.width,this.height))}e.prototype.resizePaper=function(a,b,c){var d=this,e=parseFloat(b/c),f=0,g=0;window.innerWidth?(f=window.innerWidth<b||ie8?window.innerWidth:b,g=f/e):document.documentElement&&document.documentElement.clientWidth&&(f=document.documentElement.clientWidth<b||ie8?document.documentElement.clientWidth:b,g=f/e),0==ie8?(a.setSize(b,c),document.querySelector("#adc_"+d.instanceId+"_wrapper img").style.height=g):document.querySelector("#adc_"+d.instanceId+"_container div").style.overflow="visible"},window.Map=e}();

resources/dynamic/Initializer.js
{%
    Dim widthMap = On(CurrentADC.PropValue("mapid").ToNumber(),"598","500","922","592","803","610","842")
    Dim heightMap = On(CurrentADC.PropValue("mapid").ToNumber(),"671","742","582","801","656","760","596")
    Dim adjustX = On(CurrentADC.PropValue("mapid").ToNumber(),"-5,5,0,0,0,0,0,0,0,-10,0,8,0","0,20,0,-5,-15,-5,0,0,20,10,10,0","0,0,50,0,0,0,0,30,0,0,0,0,0,30,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,-5,20,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0","-10,10,0,0,0,10,-5,0,0,0,0,0,20,0,0,5","10,0,0,0,0,0,-10,0,20,0,0","10,0,-10,0,0,15,10,-10,0,0,10,10,0,0,-5,0,0,0,0,0","0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0")
    Dim adjustY = On(CurrentADC.PropValue("mapid").ToNumber(),"10,0,-5,10,10,0,-30,10,0,0,0,20,0","25,10,0,-40,5,0,-10,0,10,0,30,0","0,0,0,0,0,0,0,20,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,10,0,10,0,0,-20,0,0,0,0,0,0,0,0,0,0,30,0,0,0,0,0,30","0,0,10,-10,0,5,0,-10,15,0,0,0,20,0,0,0","-30,0,0,0,0,-30,0,-10,-20,0,0","0,20,0,-15,0,0,0,0,0,0,0,-10,-15,10,10,0,0,0,0,0","0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0")
	Dim pathFileName = On(CurrentADC.PropValue("mapid").ToNumber(),"fr","uk","usa","de","be","it","ch")
%}
(function () {
    var map = new Map({
        instanceId: {%= CurrentADC.InstanceId %},
        mapId: {%:= CurrentADC.PropValue("mapid").ToNumber() %},
        width: {%:= widthMap.ToNumber() %},
        height: {%:= heightMap.ToNumber() %},
        fontSize: {%:= CurrentADC.PropValue("fontSize")%},
        animatePath: '{%:= CurrentADC.PropValue("animatePath")%}',
        fill : '{%= CurrentADC.PropValue("fill").ToHexa() %}',
        fillText: '{%= CurrentADC.PropValue("fillText").ToHexa() %}',
        opacityText: {%:= CurrentADC.PropValue("opacityText")%},
        strokeColor : '{%= CurrentADC.PropValue("strokeColor").ToHexa() %}',
        strokeWidth : {%:= CurrentADC.PropValue("strokeWidth")%},
        opacity : 1.0,
        fillHover : '{%= CurrentADC.PropValue("fillHover").ToHexa() %}',
        fillTextHover: '{%= CurrentADC.PropValue("fillTextHover").ToHexa() %}',
        opacityTextHover: {%:= CurrentADC.PropValue("opacityTextHover")%},
        strokeColorHover : '{%= CurrentADC.PropValue("strokeColorHover").ToHexa() %}',
        strokeWidthHover : {%:= CurrentADC.PropValue("strokeWidthHover")%},
        opacityHover : 1.0,
        fillSelected : '{%= CurrentADC.PropValue("fillSelected").ToHexa() %}',
        fillTextSelected: '{%= CurrentADC.PropValue("fillTextSelected").ToHexa() %}',
        opacityTextSelected: {%:= CurrentADC.PropValue("opacityTextSelected")%},
        strokeColorSelected : '{%= CurrentADC.PropValue("strokeColorSelected").ToHexa() %}',
        strokeWidthSelected : {%:= CurrentADC.PropValue("strokeWidthSelected")%},
        opacitySelected : 1.0,
        adjustXtextObj : [{%:= adjustX %}],
        adjustYtextObj : [{%:= adjustY %}],
        currentQuestion: '{%:= CurrentQuestion.Shortcut %}',
        mapCaptions : [{% Dim i
        For i = 1 To CurrentQuestion.Responses.Count
    	%}"{%:= CurrentQuestion.Responses[i].Caption.Replace("\n","\" + " + crlf + "\"") %}"{%:= On((i <> CurrentQuestion.Responses.Count),",","") %}{% Next i %}],
        mapNames : [{% Dim j
        For j = 1 To CurrentQuestion.Responses.Count
		%}'zone{%:= CurrentQuestion.Responses[j].InputValue() %}'{%:= On((i <> CurrentQuestion.Responses.Count),",","") %}{% Next j %}],
        mapPaths : [{%:= CurrentADC.GetContent("dynamic/" + pathFileName + ".txt").ToText() %}]
    });
} ());


resources/static/map.js
(function () {

	/**
	* Add event listener in DOMElement
	*
	* @param {HTMLElement} obj HTMLElement which should be listen
	* @param {String} type Type of the event to listen
	* @param {Function} fn Callback function
	*/
	function addEvent(obj, type, fn) {
        if (typeof obj.addEventListener === 'function') {
            obj.addEventListener(type, fn, false);
        } else if (obj.attachEvent) {
			obj['e' + type + fn] = fn;
			obj[type + fn] = function () {
				obj['e' + type + fn].call(obj, window.event);
			}
			obj.attachEvent('on' + type, obj[type + fn]);
		}
	}

	/**
	* Apply the hover styles to the path hovered
	*
	* @param {Object} obj Reference to the Raphael set object
	* @param {Object} opt Options which store the style to apply for the hovered path
	* @param {Object} optTxt Options which store the style to apply for the hovered text
	* @param {String} animPath Animate the path when hovered over
	*/
	function hoverIn(obj, opt, optTxt, animPath) {
		if (document.querySelector("." + this.data("name")).checked || this.data("strID") !== "area") return;
		obj.forEach(
			(function(currentName){
				return function(e) {
					if (currentName === e.data("name") && e.data("strID") === "path") {
						e.attr({
							fill: opt.fill,
							stroke: opt.stroke,
							'stroke-width': opt['stroke-width'],
							opacity: opt.opacity,
							cursor: 'pointer'
						});
						if (ie8 === false && animPath === "yes") {
							e.animate({'fill':opt.fill,'stroke': opt.stroke,'stroke-width': opt['stroke-width'],'opacity': opt.opacity, 'transform':'s1.03 1.03'}, 200);
						}
					} else if (currentName === e.data("name") && e.data("strID") === "txt") {
						e.attr({
							fill: optTxt.fill,
							stroke: optTxt.stroke,
							'stroke-width': optTxt['stroke-width'],
							opacity: optTxt.opacity,
							cursor: 'pointer'
						});
					}
				};
			})(this.data("name"))
		);
	}

	/**
	* Put back the original styles to the path hovered out
	* @param {Object} obj Reference to the Raphael set object
	* @param {Object} opt Options which store the style to apply for the path
	* @param {Object} optTxt Options which store the style to apply for the text
	* @param {String} animPath Animate the path
	*/
	function hoverOut(obj, opt, optTxt, animPath) {
		if (document.querySelector("." + this.data("name")).checked || this.data("strID") !== "area") return;
		obj.forEach(
			(function(currentName){
				return function(e) {
					if (currentName === e.data("name") && e.data("strID") === "path") {
						if (ie8 === false && animPath === "yes") {
							e.animate({'fill':opt.fill,'stroke': opt.stroke,'stroke-width': opt['stroke-width'],'opacity': opt.opacity, 'transform':"s1.0 1.0"}, 200);
						}
						e.attr({
							fill: opt.fill,
							stroke: opt.stroke,
							'stroke-width': opt['stroke-width'],
							opacity: opt.opacity,
							cursor: 'pointer'
						});
					} else if (currentName === e.data("name") && e.data("strID") === "txt") {
						e.attr({
							fill: optTxt.fill,
							stroke: optTxt.stroke,
							'stroke-width': optTxt['stroke-width'],
							opacity: optTxt.opacity,
							cursor: 'pointer'
						});
					}
				};
			})(this.data("name"))
		);
	}

	/**
	* Select the input when clicking on a path and apply the styles
	*
	* @param {Object} obj Reference to the Raphael set object
	* @param {Object} opt Options which store the style to apply for the selected path
	* @param {Object} optTxt Options which store the style to apply for the selected text
	* @param {Object} optOff Options which store the style to apply for the non selected path
	* @param {Object} optTxtOff Options which store the style to apply for the non selected text
	* @param {String} animPath Animate the path
	*/
	function clickZone(obj,opt, optTxt, optOff, optTxtOff, animPath, currentQuestion) {
        if (this.data("strID") !== "area") return;
		document.querySelector("." + this.data("name")).checked = true;
		obj.forEach(
			(function(currentName){
				return function(e) {
					if (currentName !== e.data("name") && e.data("strID") === "path") {
						e.attr({
							fill: optOff.fill,
							stroke: optOff.stroke,
							'stroke-width': optOff['stroke-width'],
							opacity: optOff.opacity,
							cursor: 'pointer'
						});
					} else if (currentName !== e.data("name") && e.data("strID") === "txt") {
						e.attr({
							fill: optTxtOff.fill,
							stroke: optTxtOff.stroke,
							'stroke-width': optTxtOff['stroke-width'],
							opacity: optTxtOff.opacity,
							cursor: 'pointer'
						});
					} else if (currentName === e.data("name") && e.data("strID") === "path") {
						e.attr({
							fill: opt.fill,
							stroke: opt.stroke,
							'stroke-width': opt['stroke-width'],
							opacity: opt.opacity,
							cursor: 'pointer'
						});
						if (ie8 === false && animPath === "yes") {
							e.animate({'fill':opt.fill,'stroke': opt.stroke,'stroke-width': opt['stroke-width'],'opacity': opt.opacity, 'transform':"s1.0 1.0"}, 200);
						}
					} else if (currentName === e.data("name") && e.data("strID") === "txt") {
						e.attr({
							fill: optTxt.fill,
							stroke: optTxt.stroke,
							'stroke-width': optTxt['stroke-width'],
							opacity: optTxt.opacity,
							cursor: 'pointer'
						});
					}
				};
			})(this.data("name"))
		);
        if (window.askia 
            && window.arrLiveRoutingShortcut 
            && window.arrLiveRoutingShortcut.length > 0
            && window.arrLiveRoutingShortcut.indexOf(currentQuestion) >= 0) {
            askia.triggerAnswer();
        }
	}

	/**
	* Creates a new instance of the map
	*
	* @param {Object} options Options of the map
	* @param {String} options.instanceId=1 Id of the ADC instance
    * @param {Number} options.mapId=1 Id of the map
	* @param {Number} options.width=300 Width of the image/ViewBox
	* @param {Number} options.height=300 Height of the image/ViewBox
	* @param {Number} options.fontSize=16 Font size of the text
	* @param {String} options.animatePath="yes" Animate the path when hovered over
	* @param {String} options.fill="#F7F7F7" Background color of the path
	* @param {String} options.strokeColor="#A1A7AD" Border color of the path
	* @param {Number} options.strokeWidth=1 Border Thickness of the path
	* @param {Number} options.opacity=0.0 Opacity of the path
	* @param {String} options.fillHover="#FFFFFF" Background color of the path when hovered
	* @param {String} options.strokeColorHover="#A1A7AD" Border color of the path when hovered
	* @param {Number} options.strokeWidthHover=1 Border Thickness of the path when hovered
	* @param {Number} options.opacityHover=0.0 Opacity of the path when hovered
	* @param {String} options.fillSelected="#E8F1FF" Background color of the path when selected
	* @param {String} options.strokeColorSelected="#A1A7AD" Border color of the path when selected
	* @param {Number} options.strokeWidthSelected=1 Border Thickness of the path when selected
	* @param {Number} options.opacitySelected=0.0 Opacity of the path when selected
	* @param {String} options.fillText="#444444" Font color of the text
	* @param {Number} options.opacityText=0.0 Opacity of the text
	* @param {String} options.fillTextHover="#444444" Font color of the text when hovered over
	* @param {Number} options.opacityTextHover=0.0 Opacity of the text when hovered over
	* @param {String} options.fillTextSelected="#444444" Font color of the text when selected
	* @param {Number} options.opacityTextSelected=0.0 Opacity of the text when selected
	* @param {Array} options.mapCaptions=[] Array of caption for the area
	* @param {Array} options.mapNames=[] Array of name for the area
	* @param {Array} options.mapPaths=[] Array of path for the area
	*/
	function Map(options) {
		this.options = options;
		this.instanceId = options.instanceId || 1;
        this.mapId = options.mapId || 1;
		this.width = options.width || 300;
		this.height = options.height || 300;
		this.fontSize = options.fontSize || 16;
		this.animatePath = options.animatePath || "yes";
		// Path
		this.fill = options.fill ||  "#F7F7F7";
		this.strokeColor = options.strokeColor || "#A1A7AD";
		this.strokeWidth = options.strokeWidth || 1;
		this.opacity = options.opacity || 0.0;
		this.fillHover = options.fillHover ||  "#FFFFFF";
		this.strokeColorHover = options.strokeColorHover || "#A1A7AD";
		this.strokeWidthHover = options.strokeWidthHover || 1;
		this.opacityHover = options.opacityHover || 0.0;
		this.fillSelected = options.fillSelected ||  "#E8F1FF";
		this.strokeColorSelected = options.strokeColorSelected || "#A1A7AD";
		this.strokeWidthSelected = options.strokeWidthSelected || 1;
		this.opacitySelected = options.opacitySelected || 0.0;
		// Text
		this.fillText = options.fillText || "#444444";
		this.opacityText = options.opacityText || 0.0;
		this.fillTextHover = options.fillTextHover || "#444444";
		this.opacityTextHover = options.opacityTextHover || 0.0;
		this.fillTextSelected = options.fillTextSelected || "#007BC6";
		this.opacityTextSelected = options.opacityTextSelected || 0.0;
		this.adjustXtextObj = options.adjustXtextObj || [];
		this.adjustYtextObj = options.adjustYtextObj || [];
		this.mapCaptions = options.mapCaptions || [];
		this.mapNames = options.mapNames || [];
		this.mapPaths = options.mapPaths || [];
        var currentQuestion = options.currentQuestion || "";

		var animPath = this.animatePath;

		var optArea = {
			fill: this.fill,
			stroke: this.strokeColor,
			'stroke-width': this.strokeWidth,
			opacity: 0,
			cursor: 'pointer'
		}
		var optPath = {
			'stroke-linecap': 'round',
			'stroke-linejoin': 'round',
			fill: this.fill,
			stroke: this.strokeColor,
			'stroke-width': this.strokeWidth,
			opacity: this.opacity,
			cursor: 'pointer'
		}
		var optPathHover = {
			'stroke-linecap': 'round',
			'stroke-linejoin': 'round',
			fill:this.fillHover,
			stroke:this.strokeColorHover,
			'stroke-width': this.strokeWidthHover,
			opacity: this.opacityHover
		};
		var optPathSelected = {
			'stroke-linecap': 'round',
			'stroke-linejoin': 'round',
			fill:this.fillSelected,
			stroke:this.strokeColorSelected,
			'stroke-width': this.strokeWidthSelected,
			opacity: this.opacitySelected
		};
		var optText = {
			'font-size': this.fontSize,
			fill: this.fillText,
			stroke: 'none',
			'stroke-width': 0,
			opacity: this.opacityText,
			'font-weight': 'normal',
			cursor: 'pointer'
		}
		var optTextHover = {
			'font-size': this.fontSize,
			fill: this.fillTextHover,
			stroke: 'none',
			'stroke-width': 0,
			opacity: this.opacityTextHover,
			'font-weight': 'normal'
		}
		var optTextSelected = {
			'font-size': this.fontSize,
			fill: this.fillTextSelected,
			stroke: 'none',
			'stroke-width': 0,
			opacity: this.opacityTextSelected,
			'font-weight': 'normal'
		}

		var paper = Raphael("adc_" + this.instanceId + "_container");
		var set = paper.set();
		
		for(var i = 0;l = this.mapPaths.length, i < l; i++){
			var path = paper.path(this.mapPaths[i]).attr(document.querySelector("." + this.mapNames[i]).checked ? optPathSelected : optPath).data('name',this.mapNames[i]).data('strID','path');			
			set.push(path);
		}
		for (var i = 0;l = this.mapCaptions.length, i < l; i++) {
			var bBox = set[i].getBBox();
			var textObj = path.paper.text( (bBox.x + bBox.width / 2) + ((this.adjustXtextObj[i] === undefined) ? 0 : this.adjustXtextObj[i]), (bBox.y + bBox.height / 2) + ((this.adjustYtextObj[i] === undefined) ? 0 : this.adjustYtextObj[i]), this.mapCaptions[i].replace(/\r\n/g, "\n") ).attr( document.querySelector("." + this.mapNames[i]).checked ? optTextSelected : optText ).data('name',this.mapNames[i]).data('strID','txt');
			set.push(textObj);
		}
		for (var i = 0;l = this.mapPaths.length, i < l; i++) {
			var area = paper.path(this.mapPaths[i]).attr(optArea).data('name',this.mapNames[i]).data('strID','area');
			set.push(area);
		}

		set.hover(function () {
			hoverIn.call(this, set, optPathHover, optTextHover, animPath)
		}, function () {
			hoverOut.call(this, set, optPath, optText, animPath)
		});

		set.click(function () {
			clickZone.call(this, set, optPathSelected, optTextSelected, optPath, optText, animPath, currentQuestion)
		});

		paper.setViewBox(0, 0, this.width, this.height, false);

		this.resizePaper(paper, this.width, this.height);
		addEvent(document.getElementsByTagName("BODY")[0], "resize", this.resizePaper(paper, this.width, this.height));
	}

	/**
	* Resize the map
	*
	* @param {Object} obj Reference to the Raphael paper object
	* @param {Number} width Width of the image/ViewBox
	* @param {Number} height Height of the image/ViewBox
	*/
	Map.prototype.resizePaper = function (obj, width, height) {
		var self = this;
		var ratio = parseFloat(width / height);
		var w = 0, h = 0;
		if (window.innerWidth){
			w = (window.innerWidth < width || ie8) ? window.innerWidth : width;
			h = w / ratio;
		} else if (document.documentElement && document.documentElement.clientWidth) {
			w = (document.documentElement.clientWidth < width || ie8) ? document.documentElement.clientWidth : width;
			h = w / ratio;
		}

		if (ie8 == false) {
			// any code here will be executed by IE 9 and greater
			obj.setSize(width, height);
			document.querySelector("#adc_" + self.instanceId + "_wrapper img").style.height = h;
		} else {
			// any code here will be executed by IE 8 and lower
			document.querySelector("#adc_" + self.instanceId + "_container div").style.overflow = 'visible';
		}
	};

	/**
	* Attach the Map to the window object
	*/
	window.Map = Map;
}());
